src/Algebra/Graph/NonEmpty.hs

```diff
@@ -0,0 +1,563 @@
+-- edge x y               == 'connect' ('vertex' x) ('vertex' y)
+-- 'hasEdge' x y (edge x y) == True
+-- 'edgeCount'   (edge x y) == 1
+-- 'vertexCount' (edge 1 1) == 1
+-- 'vertexCount' (edge 1 2) == 2
+-- @
+edge :: a -> a -> NonEmptyGraph a
+edge u v = connect (vertex u) (vertex v)
+
+-- | /Overlay/ two graphs. An alias for the constructor 'Overlay'. This is an
+-- idempotent, commutative and associative operation with the identity 'empty'.
+-- Complexity: /O(1)/ time and memory, /O(s1 + s2)/ size.
+--
+-- @
+-- 'hasVertex' z (overlay x y) == 'hasVertex' z x || 'hasVertex' z y
+-- 'vertexCount' (overlay x y) >= 'vertexCount' x
+-- 'vertexCount' (overlay x y) <= 'vertexCount' x + 'vertexCount' y
+-- 'edgeCount'   (overlay x y) >= 'edgeCount' x
+-- 'edgeCount'   (overlay x y) <= 'edgeCount' x   + 'edgeCount' y
+-- 'size'        (overlay x y) == 'size' x        + 'size' y
+-- 'vertexCount' (overlay 1 2) == 2
+-- 'edgeCount'   (overlay 1 2) == 0
+-- @
+overlay :: NonEmptyGraph a -> NonEmptyGraph a -> NonEmptyGraph a
+overlay = Overlay
+
+-- | /Connect/ two graphs. An alias for the constructor 'Connect'. This is an
+-- associative operation with the identity 'empty', which distributes over the
+-- overlay and obeys the decomposition axiom.
+-- Complexity: /O(1)/ time and memory, /O(s1 + s2)/ size. Note that the number
+-- of edges in the resulting graph is quadratic with respect to the number of
+-- vertices of the arguments: /m = O(m1 + m2 + n1 * n2)/.
+--
+-- @
+-- 'hasVertex' z (connect x y) == 'hasVertex' z x || 'hasVertex' z y
+-- 'vertexCount' (connect x y) >= 'vertexCount' x
+-- 'vertexCount' (connect x y) <= 'vertexCount' x + 'vertexCount' y
+-- 'edgeCount'   (connect x y) >= 'edgeCount' x
+-- 'edgeCount'   (connect x y) >= 'edgeCount' y
+-- 'edgeCount'   (connect x y) >= 'vertexCount' x * 'vertexCount' y
+-- 'edgeCount'   (connect x y) <= 'vertexCount' x * 'vertexCount' y + 'edgeCount' x + 'edgeCount' y
+-- 'size'        (connect x y) == 'size' x        + 'size' y
+-- 'vertexCount' (connect 1 2) == 2
+-- 'edgeCount'   (connect 1 2) == 1
+-- @
+connect :: NonEmptyGraph a -> NonEmptyGraph a -> NonEmptyGraph a
+connect = Connect
+
+-- | Construct the graph comprising a given list of isolated vertices.
+-- Complexity: /O(L)/ time, memory and size, where /L/ is the length of the
+-- given list.
+--
+-- @
+-- vertices1 [x]           == 'vertex' x
+-- 'hasVertex' x . vertices1 == 'elem' x
+-- 'vertexCount' . vertices1 == 'length' . 'Data.List.NonEmpty.nub'
+-- 'vertexSet'   . vertices1 == Set.'Set.fromList' . 'Data.List.NonEmpty.toList'
+-- @
+vertices1 :: NonEmpty a -> NonEmptyGraph a
+vertices1 = overlays1 . fmap vertex
+
+-- | Construct the graph from a list of edges.
+-- Complexity: /O(L)/ time, memory and size, where /L/ is the length of the
+-- given list.
+--
+-- @
+-- edges1 [(x,y)]     == 'edge' x y
+-- 'edgeCount' . edges1 == 'Data.List.NonEmpty.length' . 'Data.List.NonEmpty.nub'
+-- @
+edges1 :: NonEmpty (a, a) -> NonEmptyGraph a
```

Comment:
side comment: should non-empty variants be generalised to `Foldable f => f (a, a) -> Graph a` ?

Reply:
Having plain types like `[a] -> Graph a` seems friendlier for newcomers.

I was thinking to make type signatures fully polymorphic in modules `Algebra.Graph.Class` and `Algebra.Graph.HigherKinded.Class` instead: these modules are a bit more advanced anyway, so their users are probably comfortable with types like `(Foldable f, Graph g) => f a -> g a`.

This allows you to `import Algebra.Graph.HigherKinded.Class (vertices)` and get a fully polymorphic version of `vertices` in scope. Does this sound like good idea?